### PR TITLE
change interface

### PR DESCRIPTION
**Pull request by miiu96**
Change index interface

### CODE REVIEW

File: elasticProcessor.go [hunk 1]
```diff
@@ -529,24 +548,32 @@ func (ei *elasticProcessor) indexAlteredAccounts(blockTimestamp uint64, accounts
 		return nil
 	}
 
-	return ei.SaveAccounts(blockTimestamp, accountsToIndex)
+	accountsEGLD := make([]*types.AccountEGLD, len(accountsToIndex))
```
Comment (bogdan-rosianu) [inline elasticProcessor.go:551]: please rename

File: types/account.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package types
```
Comment (bogdan-rosianu) [inline types/account.go:1]: suggestion to rename the package to `data`, just the way we have in `elrond-go` and other projects

File: types/account.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+type AccountBalanceHistory struct {
+	Address         string        `json:"address"`
+	Timestamp       time.Duration `json:"timestamp"`
+	Balance         string        `json:"balance"`
+	TokenIdentifier string        `json:"token,omitempty"`
+	IsSender        bool          `json:"isSender,omitempty"`
+}
+
+// AccountEGLD is a structure that is needed for EGLD accounts
+type AccountEGLD struct {
```
Comment (bogdan-rosianu) [inline types/account.go:30]: You can drop the `EGLD` suffix.

File: workItems/workItemAccounts.go [hunk 4]
```diff
@@ -23,7 +26,15 @@ func NewItemAccounts(
 
 // Save will save information about an account
 func (wiv *itemAccounts) Save() error {
-	err := wiv.indexer.SaveAccounts(wiv.blockTimestamp, wiv.accounts)
+	accountsEGLD := make([]*types.AccountEGLD, len(wiv.accounts))
```
Comment (bogdan-rosianu) [inline workItems/workItemAccounts.go:29]: accounts*

File: workItems/workItemBlock.go [hunk 5]
```diff
@@ -7,82 +7,69 @@ import (
...
 func (wib *itemBlock) Save() error {
-	if check.IfNil(wib.headerHandler) {
+	if check.IfNil(wib.argsSaveBlock.Header) {
 		log.Warn("nil header provided when trying to index block, will skip")
 		return nil
 	}
 
-	log.Debug("indexer: starting indexing block", "hash", wib.headerHash, "nonce", wib.headerHandler.GetNonce())
+	log.Debug("indexer: starting indexing block",
+		"hash", logger.DisplayByteSlice(wib.argsSaveBlock.HeaderHash),
```
Comment (bogdan-rosianu) [inline workItems/workItemBlock.go:43]: why do you need `logger.DisplayByteSlice`? I think the hash was displayed correctly without it

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: elasticProcessor.go [hunk 6]
```diff
@@ -318,6 +319,12 @@ func (ei *elasticProcessor) RemoveMiniblocks(header data.HeaderHandler, body *bl
 	return ei.elasticClient.DoBulkRemove(miniblocksIndex, encodedMiniblocksHashes)
 }
 
+// RemoveTransactions will remove transaction that are in miniblock from the elasticsearch server
```
[the remark this answers is not in the document]
Comment (miiu96) [reply elasticProcessor.go:322]: done.

File: elasticProcessor.go (hunk 1 above)
Comment (miiu96) [reply elasticProcessor.go:551]: done.

File: types/account.go (hunk 2 above)
Comment (miiu96) [reply types/account.go:1]: done.

File: types/account.go (hunk 3 above)
Comment (miiu96) [reply types/account.go:30]: done.

File: workItems/workItemAccounts.go (hunk 4 above)
Comment (miiu96) [reply workItems/workItemAccounts.go:29]: renamed.

File: workItems/workItemBlock.go (hunk 5 above)
Comment (miiu96) [reply workItems/workItemBlock.go:43]: removed.

File: data/buffer.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
+package data
+
+import "bytes"
+
+// BulkSizeThreshold is the constant for the maximum size of one bulk request that is sent to the elaticsearch database
+const BulkSizeThreshold = 800000 // 0.8MB
```
Comment (iulianpascalau) [inline data/buffer.go:6]: This is not quite 0.8MB :D

File: data/buffer.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+// BufferSlice extend structure bytes.Buffer with new methods
+type bufferSlice struct {
```
Comment (iulianpascalau) [inline data/buffer.go:9]: unit tests?

File: data/transaction.go [hunk 9]
```diff
@@ -34,7 +32,7 @@ type Transaction struct {
 	SenderUserName       []byte        `json:"senderUsername,omitempty"`
 	ReceiverUserName     []byte        `json:"receiverUsername,omitempty"`
 	Log                  TxLog         `json:"-"`
-	rcvAddrBytes         []byte
+	RcvAddrBytes         []byte        `json:"-"`
```
Comment (iulianpascalau) [inline data/transaction.go:35]: ReceiverAddressBytes ? (+ rename the according getter?)

File: dataIndexer.go [hunk 10]
```diff
@@ -124,17 +114,36 @@ func (di *dataIndexer) RevertIndexedBlock(header data.HeaderHandler, body data.B
 }
 
 // SaveRoundsInfo will save data about a slice of rounds in elasticsearch
-func (di *dataIndexer) SaveRoundsInfo(roundsInfo []workItems.RoundInfo) {
-	wi := workItems.NewItemRounds(di.elasticProcessor, roundsInfo)
+func (di *dataIndexer) SaveRoundsInfo(roundsInfo []*indexer.RoundInfo) {
+	roundsInfoE := make([]*data.RoundInfo, 0)
+	for _, info := range roundsInfo {
+		roundsInfoE = append(roundsInfoE, &data.RoundInfo{
```
Comment (iulianpascalau) [inline dataIndexer.go:120]: roundsInfo *?

File: elasticProcessor.go [hunk 11]
```diff
@@ -341,16 +348,28 @@ func (ei *elasticProcessor) SaveMiniblocks(header data.HeaderHandler, body *bloc
...
 	}
 
-	txs, alteredAccounts := ei.prepareTransactionsForDatabase(body, header, txPool, selfShardID)
+	allTxs := make(map[string]nodeData.TransactionHandler)
+
+	sliceMaps := []map[string]nodeData.TransactionHandler{
+		pool.Txs, pool.Scrs, pool.Receipts, pool.Invalid, pool.Rewards,
+	}
+
+	for _, txsMap := range sliceMaps {
```
Comment (iulianpascalau) [inline elasticProcessor.go:365]: This could have been another function "mergeSliceOfMaps"

File: elasticProcessor.go [hunk 12]
```diff
@@ -529,24 +548,32 @@ func (ei *elasticProcessor) indexAlteredAccounts(blockTimestamp uint64, accounts
...
+			IsSender:    false,
+		}
+	}
+
+	return ei.SaveAccounts(blockTimestamp, accts)
 }
 
 // SaveAccounts will prepare and save information about provided accounts in elasticsearch server
-func (ei *elasticProcessor) SaveAccounts(blockTimestamp uint64, accounts []state.UserAccountHandler) error {
+func (ei *elasticProcessor) SaveAccounts(blockTimestamp uint64, accts []*data.Account) error {
```
Comment (iulianpascalau) [inline elasticProcessor.go:563]: accountsSlice instead of accts ?

File: workItems/workItemAccounts.go [hunk 13]
```diff
@@ -23,7 +26,15 @@ func NewItemAccounts(
 
 // Save will save information about an account
 func (wiv *itemAccounts) Save() error {
-	err := wiv.indexer.SaveAccounts(wiv.blockTimestamp, wiv.accounts)
+	accountsEGLD := make([]*data.Account, len(wiv.accounts))
```
Comment (iulianpascalau) [inline workItems/workItemAccounts.go:29]: can we drop the EGLD suffix?

File: data/buffer.go (hunk 7 above)
Comment (miiu96) [reply data/buffer.go:6]: changed

File: data/transaction.go (hunk 9 above)
Comment (miiu96) [reply data/transaction.go:35]: done.

File: dataIndexer.go (hunk 10 above)
Comment (miiu96) [reply dataIndexer.go:120]: done.

File: elasticProcessor.go (hunk 11 above)
Comment (miiu96) [reply elasticProcessor.go:365]: done.

File: elasticProcessor.go (hunk 12 above)
Comment (miiu96) [reply elasticProcessor.go:563]: done.

File: workItems/workItemAccounts.go (hunk 13 above)
Comment (miiu96) [reply workItems/workItemAccounts.go:29]: done.

File: data/buffer_test.go [hunk 14]
```diff
@@ -0,0 +1 @@
+package data
```
Comment (iulianpascalau) [inline data/buffer_test.go:1]: missing tests

File: data/transaction.go [hunk 15]
```diff
@@ -52,7 +52,7 @@ func (t *Transaction) GetData() []byte {
 
 // GetRcvAddr will return transaction receiver address
 func (t *Transaction) GetRcvAddr() []byte {
```
Comment (iulianpascalau) [inline data/transaction.go:54]: GetReceiverAddress ?

File: interface.go [hunk 16]
```diff
@@ -24,16 +25,17 @@ type DispatcherHandler interface {
...
+	RemoveMiniblocks(header nodeData.HeaderHandler, body *block.Body) error
+	RemoveTransactions(header nodeData.HeaderHandler, body *block.Body) error
+	SaveMiniblocks(header nodeData.HeaderHandler, body *block.Body) (map[string]bool, error)
+	SaveTransactions(body *block.Body, header nodeData.HeaderHandler, pool *indexer.Pool, mbsInDb map[string]bool) error
+	SaveValidatorsRating(index string, validatorsRatingInfo []*data.ValidatorRatingInfo) error
+	SaveRoundsInfo(infos []*data.RoundInfo) error
 	SaveShardValidatorsPubKeys(shardID, epoch uint32, shardValidatorsPubKeys [][]byte) error
 	SetTxLogsProcessor(txLogsProc process.TransactionLogProcessorDatabase)
-	SaveAccounts(blockTimestamp uint64, accounts []state.UserAccountHandler) error
+	SaveAccounts(blockTimestamp uint64, accountsEGLD []*data.Account) error
```
Comment (bogdan-rosianu) [inline interface.go:38]: accounts*

File: data/account.go [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+// AccountBalanceHistory represents an entry in the user accounts balances history
+type AccountBalanceHistory struct {
+	Address         string        `json:"address"`
+	Timestamp       time.Duration `json:"timestamp"`
+	Balance         string        `json:"balance"`
+	TokenIdentifier string        `json:"token,omitempty"`
+	IsSender        bool          `json:"isSender,omitempty"`
+}
+
+// Account is a structure that is needed for EGLD accounts
```
Comment (bogdan-rosianu) [inline data/account.go:29]: regular* instead of EGLD*

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: interface.go (hunk 16 above)
Comment (miiu96) [reply interface.go:38]: done.

File: data/account.go (hunk 17 above)
Comment (miiu96) [reply data/account.go:29]: done.

**Review by bogdan-rosianu — APPROVED** [review]

File: data/transaction.go (hunk 15 above)
Comment (miiu96) [reply data/transaction.go:54]: I have to keep this name in order to implement the interface  `TransactionWithFeeHandler`

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**